ciongke/cmd/hook/main.go

```diff
@@ -0,0 +1,337 @@
+)
+
+var (
+	port         = flag.Int("port", 8888, "Port to listen on.")
+	jobImage     = flag.String("test-pr-image", "", "Image to create to test PRs.")
+	jobBucket    = flag.String("results-bucket", "", "Public GCS bucket to store results.")
+	sourceBucket = flag.String("source-bucket", "", "Private GCS bucket to store PR source code.")
+	namespace    = flag.String("namespace", "default", "Namespace for all CI objects.")
+	org          = flag.String("org", "", "GitHub org to trust.")
+	teamID       = flag.Int("team-id", 0, "GitHub team ID to trust.")
+	teamName     = flag.String("team-name", "", "GitHub team name that matches the ID.")
+
+	webhookSecretFile = flag.String("hmac-secret-file", "/etc/hmac/hmac", "Path to the file containing the GitHub HMAC secret.")
+	oauthSecretFile   = flag.String("oauth-secret-file", "/etc/oauth/oauth", "Path to the file containing the GitHub OAuth secret.")
+)
+
+var (
+	jobDeadlineSeconds int64 = 36000
+	jobResourceCPU     int64 = 1
+	jobResourceMemory  int64 = 1 * 1000 * 1000 * 1000
+)
+
+// Server implements http.Handler.
+type Server struct {
+	Events chan Event
+
+	Port         int
+	Org          string
+	TeamID       int
+	TeamName     string
+	GitHubClient *github.Client
+	HMACSecret   []byte
+
+	KubeClient      client.Interface
+	KubeBatchClient client.BatchInterface
+	Namespace       string
+}
+
+// Event is simply the GitHub event type and the JSON payload.
+type Event struct {
+	Type    string
+	Payload []byte
+}
+
+func main() {
+	flag.Parse()
+
+	webhookSecret, err := ioutil.ReadFile(*webhookSecretFile)
+	if err != nil {
+		log.Fatalf("Could not read webhook secret file: %s", err)
+	} else if len(webhookSecret) == 0 {
+		log.Fatalf("Webhook secret must not be empty.")
+	}
+
+	oauthSecret, err := ioutil.ReadFile(*oauthSecretFile)
+	if err != nil {
+		log.Fatalf("Could not read oauth secret file: %s", err)
+	} else if len(oauthSecret) == 0 {
+		log.Fatalf("OAuth secret must not be empty.")
+	}
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(oauthSecret)})
+	tc := oauth2.NewClient(oauth2.NoContext, ts)
+	githubClient := github.NewClient(tc)
+
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		log.Fatalf("Error getting rest client: %s", err)
+	}
+	kubeClient, err := client.New(config)
+	if err != nil {
+		log.Fatalf("Error getting client: %s", err)
+	}
+	kubeBatchClient, err := client.NewBatch(config)
+	if err != nil {
+		log.Fatalf("Error getting batch client: %s", err)
+	}
+
+	s := &Server{
+		Events: make(chan Event),
+
+		Port:         *port,
+		Org:          *org,
+		TeamID:       *teamID,
+		TeamName:     *teamName,
+		GitHubClient: githubClient,
+		HMACSecret:   bytes.TrimSpace(webhookSecret),
+
+		KubeClient:      kubeClient,
+		KubeBatchClient: kubeBatchClient,
+		Namespace:       *namespace,
+	}
+	go func() {
+		for event := range s.Events {
+			go s.handleEvent(event)
+		}
+	}()
+	if err := http.ListenAndServe(":"+strconv.Itoa(*port), s); err != nil {
```

Comment:
cute trick: `log.Fatal(http.ListenAndServe` -- since this exiting exits your program anyways.